src/main/java/org/spongepowered/common/service/user/UserDiscoverer.java

```diff
@@ -131,8 +164,9 @@ static User findByProfile(org.spongepowered.api.profile.GameProfile profile) {
         return user;
     }
 
+    @Nullable
     static User findByUsername(String username) {
-        User user = userByNameCache.getIfPresent(username);
+        User user = userByNameCache.getIfPresent(username.toLowerCase());
```

Comment:
Why ?
You seem to lowercase the player name, but Dualspiral and dualspiral are not the same GameProfile.

Reply:
I get that, but it's because the checks that we perform from the `UserStorageService` are case insensitive, as per the javadocs.

https://github.com/SpongePowered/SpongeAPI/blob/stable-7/src/main/java/org/spongepowered/api/service/user/UserStorageService.java#L50-L51

The key to the cache is effectively used as a search index, so this allows for faster matching - we don't have to iterate over all keys and do a case insensitive check, we just do hash matching.

Reply:
> The key to the cache is effectively used as a search index, so this allows for faster matching - we don't have to iterate over all keys and do a case insensitive check, we just do hash matching.
This is perfectly fine. I have no problem with that.

I'm more concerned about the case-insensitive part.
To clarify: if we try to lookup Dualspiral, the name is cached and the `User` pointing to Dualspiral is returned. Then we try to lookup dualspiral, the system will see the cached entry "Dualspiral" and return the `User` pointing to Dualspiral (which is wrong).
Is this correct ?
If yes, I think this needs more discussion unless I'm missing something.

Reply:
Yes, this is what it does, but I must stress that I'm simply making the impl conform to what our Javadocs say. The only issue may arise when you use a `GameProfile` to lookup a user, but at that point, I use the UUID anyway.

Reply:
Thinking about this some more, I'd be happy to update this to add a cache that has the case-sensitive username, and then a case insensitive cache (or multimap) that contains a list of users. I'm concerned it's a _little_ overkill, but it should alleviate the issue here that you raise.

Reply:
> Yes, this is what it does, but I must stress that I'm simply making the impl conform to what our Javadocs say. 

I can see that, my point is that the implementation was "fine" as it was. 
Even if the implementation was not respecting the API (case-insensitive lookup) I think the way to go is to fix the javadocs.
This would result in no breaking-change since the implementation was already "broken".